### DISCUSSION TITLE
Export and import of all store data (for non-admins)

### DISCUSSION THREAD

**Discussion by woutersamaey** [btcpayserver/btcpayserver#4967] [Ideas / Feature Requests]
It has come to my attention that some people are using our demo server for production use. While this is clearly not recommened, it did bring to light that we don't have a migration path for users who are not admins to get an export of their data and move over to a server of their own. Users who are using a 3rd party BTCPay host are in the same boat.

While this would be quite the challenge, I'd propose we code an export / import feature that:
- Collects all the stores, invoices, store-users, payouts, pullpayments, etc the user has access to
- Write that to XMLs or JSONs and bundle in a ZIP that people can download
- Provide an import feature where one can upload the ZIP bundle, recreating the stores, invoices, etc...

Some things we need to think about:
- The format can be JSON or XML, as long as it is user readable. Some people may want to alter or inspect the export.
- We could raw dump DB records, but this proves a challenge when importing on different versions as tables and columns change
   - A workaround could be that we include the version number in the ZIP bundle and only allow import to the exact same version, forcing people to install the exact same version and upgrade later. This would be the easiest solution and my recommendation.
- We could use the Greenfield data structures, but it wasn't designed for this. The import should in this case also be Greenfield-based. This avoids the DB change challenge, but poses entirely different challenges.
- Uploading and downloading of the potentially large files could take advantage of the file storage system we have?
- Only data the user has access to should be included, or maybe we can allow the user to select which of his stores he wants to export?
- Could we offer admins an "export all stores" option allowing full 1-to-1 cloning, including settings?
- The export process should be async, running in the background and maybe send an email to the user when an export is done and ready for download?

**Comment by pizzaserved** [comment]
any progress on this?

**Comment by xucian** [comment]
I'm also interested in this functionality. Going self-hosted is a real pain, gotta manually carry over a store's config

**Comment by ghost** [comment]
YES. we need this feature. I am using the demo version in production and many of us are. I installed my own btcpay server and damn. it turned into a full time job. It requires a lot of maintenance and configuration and powerful server. I tried to hire a part-time admin and he wanted 30k a year.
It's no wonder people are using the demo one as their production. 

I installed wordpress and xenforo in the past and it's very easy requires very little maintenance. 
They seem to of made it overly complicated. It seems like enterprise software which requires and enterprise administrator.

I love the software but we still need to depend on a third party because installing it, maintaining it is requires high level expertise. 
 i can't see small business owners using it unless they hire someone which they can't afford. 

But YES. Exporting your data and moving quickly to another server is an awesome idea. 

I've had to move 3 times already and the BTCPayServer instance eventually shuts down. 

I think at the beginning they installed it just to check it out. When they moved they let me hanging so I can to move to another instance. 

I wish you could just export your data to keep your invoices.
Really great idea I didn't think about that. 

I will have to install and run my own BTCPay Server eventually but I am not so great with the Linux command. I can only install WordPress. 

Till then I'm back on the demo as production! 😅

**Reply by pavlenex** [reply]
Can you elaborate what kind of It of  maintenance and configuration and powerful server are required?

**Reply by ghost** [reply]
In order to install you have to know how to use the sudo cli linux language. 
Depending on your setup you have have to install a reverse proxy. Then to maintain you have to go into the cli and do security updates. 
Depending on which coins you want you will need a lot of hd space. Its says in the documentation at least 25GB for bitcoin and lightning. I tried a small vps but it was slow with 2gb of ram and hd filled up fast. 30gb wasn't enough. 

So running your own instance is not just for anyone. You need to at least understand the cli, linux and other things. It's pretty much being an administrator 

i'm on shared hosting, cpanel with a wordpress site. Usually the administrators do all the linux stuff. I just uploaded the wordpress files and installed it and thats all. Sometimes i update my wordpress though the dashboard but thats it. Many of us don't have time to be a linux administrator. Thats why i prefer to just sign up to someones btcpay server. 

But like your docs say there is good and bad things when you depend on a third party host. I can't run my own host its too time consuming and hard. I'm only good at wordpress. 

That's why being able to export my whole store and quickly leave and import it into another btcpay server would be amazing. 

btcpay is still young so hopefully there will be more hosts in the future. There isn't many hosts at the moment. 

inspite of that the software is inspiring and wonderful with lots of future potential
